src/Raven.Server/Commercial/LetsEncryptUtils.cs

```diff
@@ -0,0 +1,968 @@
+            SetupInfo setupInfo,
+            LetsEncryptClient client,
+            CancellationToken token)
+        {
+            try
+            {
+                var host = (setupInfo.Domain + "." + setupInfo.RootDomain).ToLowerInvariant();
+                var wildcardHost = "*." + host;
+                if (client.TryGetCachedCertificate(wildcardHost, out var certBytes))
+                    return (null, certBytes);
+
+                var result = await client.NewOrder(new[] {wildcardHost}, token);
+
+                result.TryGetValue(host, out var challenge);
+                // we may already be authorized for this?
+                return (challenge, null);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException("Failed to receive challenge(s) information from Let's Encrypt.", e);
+            }
+        }
+
+        public class CompleteAuthorizationAndGetCertificateParameters
+        {
+            public Action OnValidationSuccessful;
+            public SetupProgressAndResult Progress;
+            public SetupInfo SetupInfo;
+            public LetsEncryptClient Client;
+            public (string Challange, LetsEncryptClient.CachedCertificateResult Cache) ChallengeResult;
+            public ServerStore ServerStore;
+            public CancellationToken Token;
+        }
+
+        public class CompleteClusterConfigurationParameters
+        {
+            public SetupProgressAndResult Progress;
+            public Action<IOperationProgress> OnProgress;
+            public SetupInfo SetupInfo;
+            public Func<string,string, Task> OnBeforeAddingNodesToCluster;
+            public Func<string, Task> AddNodeToCluster;
+            public Func<Action<IOperationProgress>, SetupProgressAndResult, X509Certificate2, Task> RegisterClientCertInOs;
+            public Func<StudioConfiguration.StudioEnvironment, Task> OnPutServerWideStudioConfigurationValues;
+            public Func<string, Task> OnWriteSettingsJsonLocally;
+            public Func<string, Task<string>> OnGetCertificatePath;
+            public Func<string, Task> RegisterClientCert;
+            public Func<CertificateDefinition, Task> PutCertificate;
+            public SetupMode SetupMode;
+            public string SettingsPath;
+            public bool CertificateValidationKeyUsages;
+            public LicenseType LicenseType;
+            public CancellationToken Token = CancellationToken.None;
+        }
+
+        public class CertificateHolder
+        {
+            public string CertificateForClients;
+            public X509Certificate2 Certificate;
+            public AsymmetricKeyEntry PrivateKey;
+        }
+
+
+        private static (byte[] CertBytes, CertificateDefinition CertificateDefinition) GenerateCertificate(CertificateHolder certificateHolder, string certificateName,
```

Comment:
This method is very specific to cert generation. Isn't it a candidate to move it to `CertificateUtil`?